test.rb

```diff
@@ -0,0 +1,406 @@
+            math_total= @hash_array[0][:maths]
+            physics_total=@hash_array[0][:physics]
+            chemistry_total=@hash_array[0][:chemistry]
+            count=0
+            for num in 1...@hash_array.length
+
+                    if  num< @hash_array.length && @hash_array[num][:student_id] == @hash_array[num-1][:student_id]
+
+                    math_total=math_total+ @hash_array[num][:maths]
+                    physics_total=physics_total+ @hash_array[num][:physics]
+                    chemistry_total=chemistry_total+ @hash_array[num][:chemistry]
+
+                  else
+                    # puts "#{num}  #{count} "
+                    math_average=(math_total)/(num-count)
+                    physics_average=(physics_total)/(num-count)
+                    chemistry_average=(chemistry_total)/(num-count)
+                    # puts "Phase #{num} - #{math_total} #{physics_total} #{chemistry_total} "
+                    htemp_groupby={ :student_id => @hash_array[num-1][:student_id], :maths => math_average, :physics => physics_average , :chemistry => chemistry_average }
+                    @hash_array_groupby.push(htemp_groupby)
+
+                    math_total= @hash_array[num][:maths]
+                    physics_total=@hash_array[num][:physics]
+                    chemistry_total=@hash_array[num][:chemistry]
+                    count=num
+
+                     # puts "Phase #{num} - #{math_average} #{physics_average} #{chemistry_average} "
+                  end
+
+            end
+            # puts "#{@hash_array.length}  #{count} "
+            math_average=(math_total)/(@hash_array.length-count)
+            physics_average=(physics_total)/(@hash_array.length-count)
+            chemistry_average=(chemistry_total)/(@hash_array.length-count)
+            htemp_groupby={ :student_id => @hash_array[@hash_array.length-1][:student_id], :maths => math_average, :physics => physics_average , :chemistry => chemistry_average }
+            @hash_array_groupby.push(htemp_groupby)
+
+
+
+          # Group by department
+          elsif(group_by.downcase=="department")
+
+             htemp_department=Hash.new
+             @hash_array= @hash_array.sort_by { |hsh| hsh[:department] }
+             math_total= @hash_array[0][:maths]
+             physics_total=@hash_array[0][:physics]
+             chemistry_total=@hash_array[0][:chemistry]
+             count=0
+
+             for num in 1...@hash_array.length
+
+                     if  num< @hash_array.length && @hash_array[num][:department] == @hash_array[num-1][:department]
+
+                     math_total=math_total+ @hash_array[num][:maths]
+                     physics_total=physics_total+ @hash_array[num][:physics]
+                     chemistry_total=chemistry_total+ @hash_array[num][:chemistry]
+
+                   else
+                     # puts "#{num}  #{count} "
+                     math_average=(math_total)/(num-count)
```

Comment:
(num-count) store it into a variable and use it, instead of making the calculations repeatedly